kratos/solving_strategies/schemes/residual_based_bossak_velocity_scheme.h

```diff
@@ -0,0 +1,805 @@
+            }
+        }
+
+        rModelPart.GetCommunicator().AssembleNonHistoricalData(NUMBER_OF_NEIGHBOUR_ELEMENTS);
+    }
+
+    static void CalculateBossakConstants(BossakConstants& rBossakConstants,
+                                         const double Alpha,
+                                         const double DeltaTime)
+    {
+        rBossakConstants.Alpha = Alpha;
+        rBossakConstants.Gamma = 0.5 - rBossakConstants.Alpha;
+        rBossakConstants.Beta = 0.25 * pow((1.00 - rBossakConstants.Alpha), 2);
+        rBossakConstants.C0 =
+            (1.0 - rBossakConstants.Alpha) / (rBossakConstants.Gamma * DeltaTime);
+        rBossakConstants.C1 =
+            DeltaTime / (rBossakConstants.Beta * rBossakConstants.Gamma);
+        rBossakConstants.C2 = 1.0 / (rBossakConstants.Gamma * DeltaTime);
+        rBossakConstants.C3 = (1.0 - rBossakConstants.Gamma) / rBossakConstants.Gamma;
+        rBossakConstants.C4 =
+            std::pow(DeltaTime, 2) * (-2.0 * rBossakConstants.Beta + 1.0) / 2.0;
+        rBossakConstants.C5 = std::pow(DeltaTime, 2) * rBossakConstants.Beta;
+        rBossakConstants.C6 = DeltaTime;
+    }
+
+    static void AssembleVariables(const std::vector<const VariableData*>& rVariables,
+                                  Communicator& rComm)
+    {
+        KRATOS_TRY;
+        for (auto p_variable_data : rVariables)
+        {
+            if (KratosComponents<Variable<array_1d<double, 3>>>::Has(
+                    p_variable_data->Name()))
+            {
+                const auto& r_variable =
+                    KratosComponents<Variable<array_1d<double, 3>>>::Get(
+                        p_variable_data->Name());
+                rComm.AssembleCurrentData(r_variable);
+            }
+            else if (KratosComponents<Variable<double>>::Has(p_variable_data->Name()))
+            {
+                const auto& r_variable =
+                    KratosComponents<Variable<double>>::Get(p_variable_data->Name());
+                rComm.AssembleCurrentData(r_variable);
+            }
+            else
+            {
+                KRATOS_ERROR << "Variable \"" << p_variable_data->Name() << "\" not found!\n";
+            }
+        }
+        KRATOS_CATCH("");
+    }
+
+    void UpdateAcceleration(ModelPart& rModelPart)
+    {
+        if (!mUpdateAcceleration)
+            return;
+
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+        const int number_of_elements = rModelPart.NumberOfElements();
+
+        auto acceleration_vars = GatherVariables(
+            rModelPart.Elements(),
+            [](const DerivativesExtension& rExtensions,
+               std::vector<const VariableData*>& rVec, ProcessInfo& rCurrentProcessInfo) {
+                rExtensions.GetSecondDerivativesVariables(rVec, rCurrentProcessInfo);
+            },
+            r_current_process_info);
+#pragma omp parallel for
+        for (int i = 0; i < number_of_elements; ++i)
```

Comment:
I think you should try looping over nodes rather than elements here. It is true that the adjoint scheme does it like this, but in the adjoint scheme we need to assemble element contributions in the update step.

Reply:
updated as discussed